pkg/handlers/generic/mutation/encryption/inject.go

```diff
@@ -0,0 +1,307 @@
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/variables"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/utils"
+	k8sClientUtil "github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/k8s/client"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pkg/handlers/generic/clusterconfig"
+)
+
+const (
+	// VariableName is the external patch variable name.
+	VariableName                        = "encryptionAtRest"
+	SecretKeyForEtcdEncryption          = "config"
+	defaultEncryptionSecretNameTemplate = "%s-encryption-config" //nolint:gosec // Does not contain hard coded credentials.
+	encryptionConfigurationOnRemote     = "/etc/kubernetes/pki/encryptionconfig.yaml"
+	apiServerEncryptionConfigArg        = "encryption-provider-config"
+)
+
+type encryptionPatchHandler struct {
+	client            ctrlclient.Client
+	keyGenerator      TokenGenerator
+	variableName      string
+	variableFieldPath []string
+}
+
+func NewPatch(client ctrlclient.Client, keyGenerator TokenGenerator) *encryptionPatchHandler {
+	return &encryptionPatchHandler{
+		client:            client,
+		keyGenerator:      keyGenerator,
+		variableName:      clusterconfig.MetaVariableName,
+		variableFieldPath: []string{VariableName},
+	}
+}
+
+func (h *encryptionPatchHandler) Mutate(
+	ctx context.Context,
+	obj *unstructured.Unstructured,
+	vars map[string]apiextensionsv1.JSON,
+	holderRef runtimehooksv1.HolderReference,
+	clusterKey ctrlclient.ObjectKey,
+	clusterGetter mutation.ClusterGetter,
+) error {
+	log := ctrl.LoggerFrom(ctx, "holderRef", holderRef)
+
+	encryptionVariable, err := variables.Get[carenv1.EncryptionAtRest](
+		vars,
+		h.variableName,
+		h.variableFieldPath...,
+	)
+	if err != nil {
+		if variables.IsNotFoundError(err) {
+			log.V(5).Info("encryption variable not defined")
+			return nil
+		}
+		return err
+	}
+
+	log = log.WithValues(
+		"variableName",
+		h.variableName,
+		"variableFieldPath",
+		h.variableFieldPath,
+		"variableValue",
+		encryptionVariable,
+	)
+
+	return patches.MutateIfApplicable(
+		obj, vars, &holderRef, selectors.ControlPlane(), log,
+		func(obj *controlplanev1.KubeadmControlPlaneTemplate) error {
+			cluster, err := clusterGetter(ctx)
+			if err != nil {
+				log.Error(err, "failed to get cluster from encryption mutation handler")
+				return err
+			}
+
+			found, err := h.DefaultEncryptionSecretExists(ctx, cluster)
+			if err != nil {
+				log.WithValues(
+					"defaultEncryptionSecret", defaultEncryptionSecretName(cluster.Name),
+				).Error(err, "failed to find default encryption configuration secret")
+				return err
+			}
+
+			// we do not rotate or override the secret keys for encryption configuration
+			if !found {
+				encConfig, err := h.generateEncryptionConfiguration(encryptionVariable.Providers)
```

Comment:
nit: I had to reread it a couple of times and though it was a typo for "etcd"
```suggestion
				encryptionConfig, err := h.generateEncryptionConfiguration(encryptionVariable.Providers)
```